### PR TITLE
Refactor: use Trollop for arg processing. Remove Cli class.

### PR DESCRIPTION
**Pull request by bazbremner**
The Cli class is a lot of code and specs around optparser. Trollop is
far more compact and we can achieve what Cli was doing in significantly
less code.

Trollop symbolises keys in the opts hash it generates, unlike our
Cli/optparser class, so some small changes to follow that.

Discussed with @mattbostock in meatspace, this is just a spike to demonstrate what one of the vcloud tools looks like minus a dedicated Cli class. I'd expect to update the other tools to follow suit and possibly look at any command line flag inconsistencies that come of it.

### CODE REVIEW

File: bin/vcloud-launch [hunk 1]
```diff
@@ -2,4 +2,32 @@
 
 require 'vcloud/launcher'
 
-Vcloud::Launcher::Cli.new(ARGV).run
+banner_text = <<EOF
+vcloud-launch takes a configuration describing a vCloud Org,
```
Comment (ajlanghorn) [inline bin/vcloud-launch]: Org > Organization?

File: lib/vcloud/launcher/launch.rb [hunk 2]
```diff
@@ -5,11 +5,11 @@ class Launch
       class MissingPreambleError < RuntimeError ; end
       class MissingConfigurationError < RuntimeError ; end
 
-      attr_reader :config, :cli_options
+      attr_reader :config, :options
 
-      def initialize(config_file, cli_options = {})
```
Comment (bazbremner) [inline lib/vcloud/launcher/launch.rb]: Just one point to raise here is that it seems a little ugly to have config_file as the first option, then all the other options as a hash. I'd rather just pass a hash in, I was focusing on kill Cli, but happy to clean this up if it makes sense to do so.

File: lib/vcloud/launcher/cli.rb [hunk 3]
```diff
@@ -1,113 +0,0 @@
...
-        end
-
-        @usage_text = opt_parser.to_s
-        begin
-          opt_parser.parse!(args)
-        rescue OptionParser::InvalidOption, OptionParser::MissingArgument => e
-          exit_error_usage(e)
-        end
-
-        exit_error_usage("must supply config_file") unless args.size == 1
```
Comment (mattbostock) [inline lib/vcloud/launcher/cli.rb]: With the changes in this PR, how are commandline arguments (as opposed to options) handled? I assume we'd see a backtrace for this particular case rather than 'must supply config_file'?

File: lib/vcloud/launcher/cli.rb [hunk 4]
```diff
@@ -1,113 +0,0 @@
...
-      end
-
-      def run
-        begin
-          launch = Vcloud::Launcher::Launch.new(@config_file, @options)
-          launch.run
-        rescue => error_msg
-          $stderr.puts(error_msg)
-          exit 1
-        end
```
Comment (mattbostock) [inline lib/vcloud/launcher/cli.rb]: Discussed this with @bazbremner IRL; the change to use Trollop would currently output a backtrace rather than an exception message. He's pushing a commit to fix.

File: bin/vcloud-launch (hunk 1 above)
Comment (bazbremner) [inline bin/vcloud-launch]: The help text was lifted unchanged from the Cli class, so I'm inclined to punt rewording elsewhere.

File: lib/vcloud/launcher/launch.rb (hunk 2 above)
Comment (mattbostock) [inline lib/vcloud/launcher/launch.rb]: I don't have a strong opinion on this as the distinction between command-line arguments and options seems OK to me. Either way, I think that should be a separate pull request.

**Comment by mattbostock** [comment]
Overall, I like this :+1: It removes a lot of boilerplate that seems unnecessary for such a common task as parsing CLI options.

These changes do remove a lot of tests, but we shouldn't need to test Trollop and the remaining tests (e.g. [check for an error](https://github.com/gds-operations/vcloud-launcher/blob/9cd13198ca9140866bff306ea84566c1a31af166/spec/vcloud/launcher/cli_spec.rb#L259-L266) when a non-existent file is passed in) are possibly not _that_ important. I'd be interested in your opinion on those tests @bazbremner.

File: lib/vcloud/launcher/cli.rb (hunk 4 above)
Comment (bazbremner) [inline lib/vcloud/launcher/cli.rb]: Addressed in fc36eb5.

File: bin/vcloud-launch (hunk 1 above)
Comment (dcarley) [inline bin/vcloud-launch]: Besides, we'd have to argue about whether it's "organiSation" or "organiZation".

Comment (rjw1) [inline bin/vcloud-launch]: there's an argument?

File: lib/vcloud/launcher/launch.rb (hunk 2 above)
Comment (dcarley) [inline lib/vcloud/launcher/launch.rb]: I quite like the distinction that `config_file` is required whereas everything else is optional.

**Comment by dcarley** [comment]
It's worth thinking about how this would apply to vcloud-configure-edge (which had some logic in the binstub and prompted us to [replace methadone](https://github.com/gds-operations/vcloud-edge_gateway/pull/70)) and vcloud-login (which has quite a lot of [tests to do with input, ouput and status codes](https://github.com/gds-operations/vcloud-core/blob/master/spec/vcloud/core/login_cli_spec.rb#L41-L146))

File: bin/vcloud-launch [hunk 5]
```diff
@@ -2,4 +2,37 @@
...
+and tries to make it a reality.
+
+See https://github.com/gds-operations/vcloud-tools for more info
+EOF
+
+opts = Trollop::options do
+  banner banner_text
+
+  opt :config_file,       "Configuration file",    required: true
```
Comment (mattbostock) [inline bin/vcloud-launch]: Discussion from meatspace: This would make the configuration file a command-line option rather than an argument as it is currently.

My preference is to keep it as an argument; that would seem more idiomatic for Unix tools. Also, the configuration's file importance as a required argument and the location where the machines to be launched are defined would suggest to me it makes sense to keep as an argument.

File: lib/vcloud/launcher/cli.rb (hunk 3 above)
Comment (mattbostock) [inline lib/vcloud/launcher/cli.rb]: Confirmed with @bazbremner IRL that Trollop will output an error since config file is defined as a required option.

File: lib/vcloud/launcher/launch.rb (hunk 2 above)
Comment (annashipman) [inline lib/vcloud/launcher/launch.rb]: :+1: to @dcarley's comment but I also agree with @mattbostock that it should be a separate PR either way

**Comment by annashipman** [comment]
I was going to comment further but then I saw @dcarley's summary with which I entirely agree so I shall back out again :smile:

File: lib/vcloud/launcher/cli.rb (hunk 3 above)
Comment (bazbremner) [inline lib/vcloud/launcher/cli.rb]: Existing behavior restored in 26cbdd3a.

**Comment by bazbremner** [comment]
@mattbostock I'm somewhat biased against the existing tests, given I removed them in favour of trusting that the library is doing the correct thing.

I see that there's value in ensuring that the code validates _somewhere_ that required information is passed in and handled appropriately.

Checking that the command-line code checks for this is ideal, but testing a script (or having a separate dedicated Cli class just feels a rather heavyweight.

In the specific case of a missing configuration file, I'd be happy provided it's explicitly handled somewhere.
Provided the user sees a minimally helpful message if the command-line code regresses. Given that loading configuration files is Vcloud::Core::ConfigLoader's purpose in life, that should be the last gasp.

File: bin/vcloud-launch (hunk 1 above)
Comment (ajlanghorn) [inline bin/vcloud-launch]: Okay, I'm fine with it being a separate PR, but ultimately it should be re-phrased around the VMware approved spelling.

**Comment by bazbremner** [comment]
@dcarley @annashipman It sounds like this change is over the first hurdle of mostly-positive review, but has raised some concerns about tests which I'll look at.

As a first attempt, I'd suggest that the general pattern should perhaps be that the script(s) use trollop to capture command line arguments and options, which are handed off to a Cli class for validation, exception classes are raised by Cli (which can then be tested for situations that warrant testing - versus situations that should be squarely in the option parser's wheelhouse).

I've never been a stickler for testing the command line wrappers heavily and instead have pushed that down to libraries. Perhaps that's coming out again and it's not an appropriate approach.

I'm interested to find a way to remove code and unnecessary tests whilst still retaining confidence that the code works along with tests that validate important behaviour.

I don't know what that balance looks like just yet; this was a spike but more feedback is certainly welcome.

**Comment by bazbremner** [comment]
@mattbostock @annashipman @dcarley Thanks for the feedback so far. I feel like I've been through the mill a little bit, which is refreshing :)

**Comment by mattbostock** [comment]
Hi @bazbremner,

Per Dan's suggestion above, have you had a chance to look at applying this change to the vCloud Edge Gateway or vCloud Login commands? I think that would be a good litmus test to know if we can apply this change across the vCloud Tools.

**Comment by bazbremner** [comment]
¡Hola @mattbostock!

I've made a small start on vcloud-edge_configure, but have been fighting tests and busy with other things. I'll see about progressing it RSN.

**Comment by mattbostock** [comment]
Thanks for all of your work on this @bazbremner :rocket: I'm going to close this pull request for now as I think it has served its purpose as a spike. We can look at this again once the work on vCloud Edge Configure is ready.
